src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -2872,40 +2882,84 @@ private int[] SerializeMethodBodies(BlobBuilder ilBuilder, PdbWriter nativePdbWr
-            bool isSmallBody = ilLength < 64 && methodBody.MaxStack <= 8 && localSignatureHandleOpt.IsNil && exceptionRegions.Length == 0;
-
-            // Check if an identical method body has already been serialized. 
+        private int SerializeThrowNullMethodBody(MethodBodyStreamEncoder encoder)
+        {
+            UserStringHandle mvidStringHandle = default(UserStringHandle);
+            Blob mvidStringFixup = default(Blob);
+
+            int offset = SerializeMethodBody(
+                encoder,
+                ThrowNullIL,
+                ThrowNullMaxStack,
+                ImmutableArray<ExceptionHandlerRegion>.Empty,
+                default(StandaloneSignatureHandle),
+                MethodBodyAttributes.None,
+                ref mvidStringHandle,
+                ref mvidStringFixup);
+
+            Debug.Assert(mvidStringHandle.Equals(default(UserStringHandle)) && mvidStringFixup.Equals(default(Blob)));
+            return offset;
+        }
+
+        private int SerializeMethodBody(
+            MethodBodyStreamEncoder encoder,
+            IMethodBody methodBody,
+            StandaloneSignatureHandle localSignatureHandleOpt,
+            ref UserStringHandle mvidStringHandle,
+            ref Blob mvidStringFixup)
+        {
+            return SerializeMethodBody(
+                encoder,
+                methodBody.IL,
+                methodBody.MaxStack,
+                methodBody.ExceptionRegions,
+                localSignatureHandleOpt,
+                (methodBody.LocalsAreZeroed ? MethodBodyAttributes.InitLocals : MethodBodyAttributes.None),
+                ref mvidStringHandle,
+                ref mvidStringFixup);
+        }
+
+        private int SerializeMethodBody(
+            MethodBodyStreamEncoder encoder,
+            ImmutableArray<byte> il,
+            ushort maxStack,
+            ImmutableArray<ExceptionHandlerRegion> exceptionRegions,
+            StandaloneSignatureHandle localSignatureHandleOpt,
+            MethodBodyAttributes methodBodyAttributes,
+            ref UserStringHandle mvidStringHandle,
+            ref Blob mvidStringFixup)
+        {
+            int ilLength = il.Length;
+            bool isSmallBody = ilLength < 64 && maxStack <= 8 && localSignatureHandleOpt.IsNil && exceptionRegions.Length == 0;
+
+            // Check if an identical method body has already been serialized.
             // If so, use the RVA of the already serialized one.
             // Note that we don't need to rewrite the fake tokens in the body before looking it up.
 
             // Don't do small body method caching during deterministic builds until this issue is fixed
             // https://github.com/dotnet/roslyn/issues/7595
             int bodyOffset;
-            if (!_deterministic && isSmallBody && _smallMethodBodies.TryGetValue(methodBody.IL, out bodyOffset))
+            if (!_deterministic && isSmallBody && _smallMethodBodies.TryGetValue(il, out bodyOffset))
```

Comment:
How confident are you that's the case? Aleksey wasn't sure we understood the root cause of this problem, in which case it is safer to apply the same workaround.